pkg/store/copr/region_cache.go

```diff
@@ -26,11 +27,233 @@ import (
+		enc.AddString("endKey", redact.Key(loc.EndKey))
+		if loc.Buckets != nil {
+			enc.AddInt("bucketCount", len(loc.Buckets.Keys))
+		}
+		return nil
+	}))
+}
+
+func keyField(name string, key []byte) zap.Field {
+	return zap.String(name, redact.Key(key))
+}
+
+func formatRanges(ranges *KeyRanges) zap.Field {
+	return zap.Object("ranges", zapcore.ObjectMarshalerFunc(func(enc zapcore.ObjectEncoder) error {
+		enc.AddInt("count", ranges.Len())
+		if ranges.Len() > 0 {
+			enc.AddString("firstStart", redact.Key(ranges.At(0).StartKey))
+			enc.AddString("lastEnd", redact.Key(ranges.At(ranges.Len()-1).EndKey))
+		}
+		return nil
+	}))
+}
+
+func validateLocationCoverage(ctx context.Context, kvRanges []tikv.KeyRange, locs []*tikv.KeyLocation) /* valid */ bool {
+	if len(kvRanges) == 0 {
+		return true
+	}
+
+	valid := true
+
+	// First, validate monotonicity of locations to catch PD corruption
+	// Continue checking all locations to report all errors
+	for i := 1; i < len(locs); i++ {
+		prev, curr := locs[i-1], locs[i]
+		if prev == nil || curr == nil {
+			continue // Skip nil check
+		}
+
+		// Check that locations are ordered (prev.StartKey <= curr.StartKey)
+		// Empty start key means beginning of key space (minimum)
+		if len(curr.StartKey) == 0 && len(prev.StartKey) > 0 {
+			// Current starts from beginning, but previous doesn't - wrong order!
+			logutil.BgLogger().Error("BatchLocateKeyRanges locations not monotonic",
+				zap.Int("locationIndex", i),
+				zap.String("issue", "current location starts from beginning but appears after non-beginning location"),
+				keyField("prevStart", prev.StartKey),
+				keyField("currStart", curr.StartKey))
+			valid = false
+			continue // Continue checking other locations
+		}
+		if len(prev.StartKey) > 0 && len(curr.StartKey) > 0 && bytes.Compare(prev.StartKey, curr.StartKey) > 0 {
+			logutil.BgLogger().Error("BatchLocateKeyRanges locations not monotonic",
+				zap.Int("locationIndex", i),
+				keyField("prevStart", prev.StartKey),
+				keyField("currStart", curr.StartKey))
+			valid = false
+			continue // Continue checking other locations
+		}
+
+		// Check for overlaps/gaps: prev.EndKey should be <= curr.StartKey
+		// Empty end key means infinity - there should be no next location
+		if len(prev.EndKey) == 0 {
+			logutil.BgLogger().Error("BatchLocateKeyRanges location extends to infinity but is not last",
+				zap.Int("locationIndex", i-1),
+				zap.Int("totalLocations", len(locs)),
+				keyField("prevStart", prev.StartKey))
+			valid = false
+			continue // Continue checking other locations
+		}
+
+		// Both keys non-empty - check for overlap
+		if len(prev.EndKey) > 0 && len(curr.StartKey) > 0 && bytes.Compare(prev.EndKey, curr.StartKey) > 0 {
+			logutil.BgLogger().Error("BatchLocateKeyRanges locations overlap",
+				zap.Int("locationIndex", i),
+				keyField("prevEnd", prev.EndKey),
+				keyField("currStart", curr.StartKey))
+			valid = false
+			// Continue checking other locations
+		}
```

Comment:
It can be a simple overlap(x, y) function which compares x.start <= x.end <= y.start <= y.end

Reply:
Current code works, but under a assumption that range start < end
A corner case make the code break is    [3, 5] [6, 1]

Reply:
Anyway, this is just the validate code, so the defect does not really matter, you can choose to modify it or not.